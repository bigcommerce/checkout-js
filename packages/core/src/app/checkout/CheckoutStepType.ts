enum CheckoutStepType {
    Billing = 'billing',
    Customer = 'customer',
    Payment = 'payment',
    Shipping = 'shipping',
}

export default CheckoutStepType;
