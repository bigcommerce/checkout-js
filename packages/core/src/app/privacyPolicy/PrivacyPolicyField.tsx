import React, { FunctionComponent, memo } from 'react';

import { TranslatedHtml } from '../locale';
import { CheckboxFormField, Fieldset } from '../ui/form';

const PrivacyPolicyCheckboxFieldLink: FunctionComponent<{ url: string }> = ({ url }) => (
    <CheckboxFormField
        labelContent={<TranslatedHtml data={{ url }} id="privacy_policy.label" />}
        name="privacyPolicy"
    />
);

const PrivacyPolicyFieldset: FunctionComponent<{ url: string }> = ({ url }) => (
    <Fieldset additionalClassName="checkout-privacy-policy">
        <PrivacyPolicyCheckboxFieldLink url={url} />
    </Fieldset>
);

export default memo(PrivacyPolicyFieldset);
