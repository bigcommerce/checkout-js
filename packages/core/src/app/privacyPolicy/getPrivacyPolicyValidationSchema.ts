import { LanguageService } from '@bigcommerce/checkout-sdk';
import { boolean, BooleanSchema, object, ObjectSchema } from 'yup';

export interface PrivacyPolicyValidatonSchemaProps {
    isRequired: boolean;
    language: LanguageService;
}

export default function getPrivacyPolicyValidationSchema({
    isRequired,
    language,
}: PrivacyPolicyValidatonSchemaProps): ObjectSchema<{ privacyPolicy?: boolean }> {
    const schemaFields: {
        privacyPolicy?: BooleanSchema;
    } = {};

    if (isRequired) {
        schemaFields.privacyPolicy = boolean().oneOf(
            [true],
            language.translate('privacy_policy.required_error'),
        );
    }

    return object(schemaFields);
}
