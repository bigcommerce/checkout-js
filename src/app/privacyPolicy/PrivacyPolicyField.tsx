import React, { memo, FunctionComponent } from 'react';

import { parseAnchor } from '../common/utility';
import { PrivacyPolicyConfig } from '../customer/Customer';
import { withLanguage, TranslatedHtml, TranslatedString, WithLanguageProps } from '../locale';
import { CheckboxFormField, Fieldset } from '../ui/form';
import { ModalHeader, ModalLink } from '../ui/modal';
import { MultiLineText } from '../ui/text';

export enum PrivacyPolicyType {
    Link = 'link',
    Text = 'text',
}

const BasePrivacyPolicyCheckboxFieldModal: FunctionComponent<WithLanguageProps & { text: string }> = ({
    language,
    text,
}) => {
    const parsedLabel = parseAnchor(language.translate('privacy_policy.label', { url: '' }));

    if (!parsedLabel) {
        return null;
    }

    const labelContent = (<>
        { parsedLabel[0] }
        <ModalLink
            body={ <MultiLineText>{ text }</MultiLineText> }
            header={
                <ModalHeader>
                    <TranslatedString id="privacy_policy.heading" />
                </ModalHeader>
            }
        >
            { parsedLabel[1] }
        </ModalLink>
        { parsedLabel[2] }
    </>);

    return (
        <CheckboxFormField
            labelContent={ labelContent }
            name="privacyPolicy"
        />
    );
};

const PrivacyPolicyCheckboxFieldModal = withLanguage(BasePrivacyPolicyCheckboxFieldModal);

const PrivacyPolicyCheckboxFieldLink: FunctionComponent<{ url: string }> = ({
    url,
}) => (
    <CheckboxFormField
        labelContent={ <TranslatedHtml data={ { url } } id="privacy_policy.label" /> }
        name="privacyPolicy"
    />
);

const PrivacyPolicyFieldset: FunctionComponent<Pick<PrivacyPolicyConfig, 'type' | 'value'>> = ({
    type,
    value,
}) => (
    <Fieldset
        additionalClassName="checkout-privacy-policy"
    >
        { type === PrivacyPolicyType.Text ?
            <PrivacyPolicyCheckboxFieldModal text={ value } /> :
            <PrivacyPolicyCheckboxFieldLink url={ value } /> }
    </Fieldset>
);

export default memo(PrivacyPolicyFieldset);
