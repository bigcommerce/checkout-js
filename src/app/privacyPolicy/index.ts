export { default as PrivacyPolicyField } from './PrivacyPolicyField';
export { default as getPrivacyPolicyValidationSchema } from './getPrivacyPolicyValidationSchema';
