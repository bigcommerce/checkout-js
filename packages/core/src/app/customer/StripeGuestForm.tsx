import { CustomerInitializeOptions, CustomerRequestOptions } from '@bigcommerce/checkout-sdk';
import { FieldProps, FormikProps, withFormik } from 'formik';
import React, { FunctionComponent, memo, ReactNode, useCallback, useEffect, useState } from 'react';

import CheckoutStepStatus from '../checkout/CheckoutStepStatus';
import { getAppliedStyles } from '../common/dom';
import { TranslatedString } from '../locale';
import { PrivacyPolicyField } from '../privacyPolicy';
import { Button, ButtonVariant } from '../ui/button';
import { BasicFormField, Fieldset, Legend } from '../ui/form';
import { LoadingOverlay } from '../ui/loading';

import { GuestFormValues } from './GuestForm';
import SubscribeField from './SubscribeField';

export interface StripeGuestFormProps {
    canSubscribe: boolean;
    checkoutButtons?: ReactNode;
    step: CheckoutStepStatus;
    continueAsGuestButtonLabelId: string;
    email?: string;
    isLoading: boolean;
    requiresMarketingConsent: boolean;
    defaultShouldSubscribe: boolean;
    privacyPolicyUrl?: string;
    onChangeEmail(email: string): void;
    onContinueAsGuest(data: GuestFormValues): void;
    deinitialize(options: CustomerRequestOptions): void;
    initialize(options: CustomerInitializeOptions): void;
    onShowLogin(): void;
}

const StripeGuestForm: FunctionComponent<StripeGuestFormProps & FormikProps<GuestFormValues>> = ({
    continueAsGuestButtonLabelId,
    isLoading,
    initialize,
    deinitialize,
    onChangeEmail,
    onShowLogin,
    onContinueAsGuest,
    canSubscribe,
    checkoutButtons,
    requiresMarketingConsent,
    privacyPolicyUrl,
    step,
}) => {

    const [continueAsAGuestButton, setContinueAsAGuestButton] = useState(true);
    const [emailValue, setEmailValue] = useState('');
    const [authentication, setAuthentication] = useState(false);
    const [isStripeLoading, setIsStripeLoading] = useState(true);
    const [isNewAuth, setIsNewAuth] = useState(false);
    const handleOnClickSubmitButton = () => {
        onContinueAsGuest({
            email: emailValue,
            shouldSubscribe: false,
        });
    };
    const setEmailCallback = useCallback((authenticated: boolean, email: string) => {
        onChangeEmail(email);
        setEmailValue(email);
        setContinueAsAGuestButton(!email);
        setAuthentication(authenticated);

        if(!authenticated){
            setIsNewAuth(true);
        }
    }, [setContinueAsAGuestButton, onChangeEmail]);

    useEffect(() => {
        if ((!step.isComplete || isNewAuth) && emailValue && authentication) {
            handleOnClickSubmitButton();
        }
    }, [emailValue, authentication, isNewAuth]);

    const handleLoading = useCallback((mounted: boolean) => {
        setIsStripeLoading(mounted);
    }, []);

    const stripeDeinitialize = () => {
        deinitialize({
            methodId: 'stripeupe',
        });
    };

    const stripeInitialize = () => {
        initialize( {
            methodId: 'stripeupe',
            stripeupe: {
                container: 'stripeupeLink',
                onEmailChange: setEmailCallback,
                isLoading: handleLoading,
                getStyles: getStripeStyles,
                gatewayId: 'stripeupe',
                methodId: 'card',
            },
        })};

    useEffect(() => {
        stripeInitialize();

        return () => stripeDeinitialize();
    }, []);

    const getStylesFromElement = (
        id: string,
        properties: string[]) => {
        const parentContainer = document.getElementById(id);

        if (parentContainer) {
            return getAppliedStyles(parentContainer, properties);
        }
 
            return undefined;
        
    };

    const containerId = 'stripe-card-component-field';

    const getStripeStyles: () => (Record<string, string> | undefined) = useCallback( () => {
        const formInput = getStylesFromElement(`${containerId}--input`, ['color', 'background-color', 'border-color', 'box-shadow']);
        const formLabel = getStylesFromElement(`${containerId}--label`, ['color']);
        const formError = getStylesFromElement(`${containerId}--error`, ['color']);

        return formLabel && formInput && formError ? {
            labelText: formLabel.color,
            fieldText: formInput.color,
            fieldPlaceholderText: formInput.color,
            fieldErrorText: formError.color,
            fieldBackground: formInput['background-color'],
            fieldInnerShadow: formInput['box-shadow'],
            fieldBorder: formInput['border-color'],
        } : undefined;
    }, [])

    const renderCheckoutThemeStylesForStripeUPE = () => {
        return (
            <div
                className="optimizedCheckout-form-input"
                id={ `${containerId}--input` }
                placeholder="1111"
            >
                <div
                    className="form-field--error"
                >
                    <div
                        className="optimizedCheckout-form-label"
                        id={ `${containerId}--error` }
                    />
                </div>
                <div
                    className="optimizedCheckout-form-label"
                    id={ `${containerId}--label` }
                />
            </div>
        );
    };

    const renderField = useCallback((fieldProps: FieldProps<boolean>) => (
        <SubscribeField
            { ...fieldProps }
            requiresMarketingConsent={ requiresMarketingConsent }
        />
    ), [
        requiresMarketingConsent,
    ]);

    const buttonText = authentication && !isNewAuth? 'customer.continue_as_stripe_customer_action' : continueAsGuestButtonLabelId;

    return (
        <>
            <div className="checkout-form">
                <LoadingOverlay
                    hideContentWhenLoading
                    isLoading={ isStripeLoading }
                >
                    <Fieldset
                        legend={ !authentication &&
                            <Legend hidden>
                                <TranslatedString id="customer.guest_customer_text"/>
                            </Legend>
                        }
                    >
                        <div className="customerEmail-container">
                            <div className="customerEmail-body">
                                <div id="stripeupeLink"/>
                                <br/>
                                { (canSubscribe || requiresMarketingConsent) && <BasicFormField
                                    name="shouldSubscribe"
                                    render={ renderField }
                                /> }

                                { privacyPolicyUrl && <PrivacyPolicyField
                                    url={ privacyPolicyUrl }
                                /> }
                            </div>

                            <div className="form-actions customerEmail-action">
                                { (!authentication || (authentication && !isNewAuth )) && <Button
                                    className="stripeCustomerEmail-button"
                                    disabled={ continueAsAGuestButton }
                                    id="stripe-checkout-customer-continue"
                                    isLoading={ isLoading }
                                    onClick={ handleOnClickSubmitButton }
                                    testId="stripe-customer-continue-as-guest-button"
                                    type="submit"
                                    variant={ ButtonVariant.Primary }
                                >
                                    <TranslatedString id={ buttonText }/>
                                </Button> }
                            </div>
                        </div>
                        {
                            !isLoading && <p>
                                <TranslatedString id="customer.login_text"/>
                                { ' ' }
                                <a
                                    data-test="customer-continue-button"
                                    id="checkout-customer-login"
                                    onClick={ onShowLogin }
                                >
                                    <TranslatedString id="customer.login_action"/>
                                </a>
                            </p>
                        }
                        { !authentication && checkoutButtons }
                    </Fieldset>
                </LoadingOverlay>
            </div>
            { renderCheckoutThemeStylesForStripeUPE() }
        </>
    );
};

export default withFormik<StripeGuestFormProps, GuestFormValues>({
    mapPropsToValues: ({
                           email = '',
                           defaultShouldSubscribe = false,
                           requiresMarketingConsent,
                       }) => ({
        email,
        shouldSubscribe: requiresMarketingConsent ? false : defaultShouldSubscribe,
        privacyPolicy: false,
    }),
    handleSubmit: (values, { props: { onContinueAsGuest } }) => {
        onContinueAsGuest(values);
    },
})(memo(StripeGuestForm));
