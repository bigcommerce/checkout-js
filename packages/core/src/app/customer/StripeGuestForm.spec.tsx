import { mount, render } from 'enzyme';
import React, { FunctionComponent } from 'react';

import CheckoutStepType from '../checkout/CheckoutStepType';
import { PrivacyPolicyField } from '../privacyPolicy';

import StripeGuestForm, { StripeGuestFormProps } from './StripeGuestForm';

describe('StripeGuestForm', () => {
    let defaultProps: StripeGuestFormProps;
    let TestComponent: FunctionComponent<Partial<StripeGuestFormProps>>;
    const handleContinueAsGuest = jest.fn();
    const dummyElement = document.createElement('div');

    beforeEach(() => {
        defaultProps = {
            canSubscribe: true,
            continueAsGuestButtonLabelId: 'customer.continue_as_guest_action',
            defaultShouldSubscribe: false,
            isLoading: false,
            onChangeEmail: jest.fn(),
            onContinueAsGuest: handleContinueAsGuest,
            deinitialize: jest.fn(),
            updateStripeLinkAuthenticated: jest.fn(),
            initialize: jest.fn(),
            onShowLogin: jest.fn(),
            requiresMarketingConsent: false,
            step: { isActive: true,
                isComplete: false,
                isEditable: false,
                isRequired: true,
                type: CheckoutStepType.Customer },
        };

        jest.mock('../common/dom', () => ({
            getAppliedStyles: () => {
                return { color: '#cccccc' };
            },
        }));
        jest.spyOn(document, 'getElementById')
            .mockReturnValue(dummyElement);

        TestComponent = props => (
            <StripeGuestForm
                { ...defaultProps }
                { ...props }
            />
        );
    });

    it('matches snapshot', () => {
        const component = render(<TestComponent />);

        expect(component).toMatchSnapshot();
    });

    it('disables "continue as guest" button when isLoading is true', () => {
        const component = mount(
            <TestComponent
                isLoading={ true }
                onContinueAsGuest={ jest.fn() }
            />
        );
        const button = component.find('[data-test="stripe-customer-continue-as-guest-button"]');

        expect(button.prop('disabled')).toBeTruthy();
    });

    it('executes a function when on click action is triggered', () => {
        const component = mount(<TestComponent />);

        const button = component.find('button');
        const onClickEvent = button.prop('onClick');

        if (onClickEvent) {
            onClickEvent(new MouseEvent('click') as any);
        }

        expect(handleContinueAsGuest).toHaveBeenCalled();
    });

    it('initializes stripeGuestForm when component mounts', () => {
        defaultProps.initialize = jest.fn((options:any) => {
            options.stripeupe.onEmailChange('cosmefulanito@cosme.mx', true);
            options.stripeupe.isLoading(true);
            options.stripeupe?.getStyles();
        })
        mount(<TestComponent { ...defaultProps } />);

        expect(defaultProps.initialize).toHaveBeenCalled();
    });

    it('deinitializes stripeGuestForm when component mounts', () => {
        const component = mount(<TestComponent { ...defaultProps } />);

        component.unmount();

        expect(defaultProps.deinitialize).toHaveBeenCalled();
    });

    it('renders form with initial values', () => {
        const component = mount(
            <TestComponent
                defaultShouldSubscribe={ true }
                email="test@bigcommerce.com"
            />
        );

        expect(component.find('input[name="shouldSubscribe"]').prop('value'))
            .toBe(true);
    });

    it('notifies when user clicks on "sign in" button', () => {
        const handleShowLogin = jest.fn();
        const component = mount(
            <TestComponent
                onShowLogin={ handleShowLogin }
            />
        );

        component.find('[data-test="customer-continue-button"]')
            .simulate('click');

        expect(handleShowLogin)
            .toHaveBeenCalled();
    });

    it('renders newsletter field if store allows newsletter subscription', () => {
        const component = mount(
            <TestComponent
                canSubscribe={ true }
            />
        );

        expect(component.exists('input[name="shouldSubscribe"]'))
            .toBe(true);
    });

    it('renders marketing consent field', () => {
        const component = mount(
            <TestComponent
                canSubscribe={ true }
                requiresMarketingConsent={ true }
            />
        );

        expect(component.exists('input[name="shouldSubscribe"]'))
            .toBe(true);
    });

    it('sets newsletter field with default value', () => {
        const Container = ({ defaultShouldSubscribe }: { defaultShouldSubscribe: boolean }) => (
            <TestComponent
                canSubscribe={ true }
                defaultShouldSubscribe={ defaultShouldSubscribe }
            />
        );

        const componentA = mount(<Container defaultShouldSubscribe={ true } />);
        const componentB = mount(<Container defaultShouldSubscribe={ false } />);

        expect(componentA.find('input[name="shouldSubscribe"]').prop('value'))
            .toBe(true);

        expect(componentB.find('input[name="shouldSubscribe"]').prop('value'))
            .toBe(false);
    });

    it('renders privacy policy field', () => {
        const component = mount(
            <TestComponent
                privacyPolicyUrl="foo"
            />
        );

        expect(component.find(PrivacyPolicyField)).toHaveLength(1);
    });

    it('does not render "sign in" button when loading', () => {
        const component = mount(
            <TestComponent
                isLoading={ true }
            />
        );

        expect(component.find('[data-test="customer-continue-button"]')).toHaveLength(0);
    });

    it('shows different action button label if another label id was provided', () => {
        const component = mount(
            <TestComponent
                continueAsGuestButtonLabelId="customer.continue"
            />
        );

        expect(component.find('[data-test="customer-continue-button"]').text()).not.toBe('Continue as guest');
    });
});
