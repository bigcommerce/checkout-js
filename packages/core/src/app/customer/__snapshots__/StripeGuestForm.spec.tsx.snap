// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StripeGuestForm matches snapshot 1`] = `
Array [
  <div
    class="checkout-form"
  >
    <div
      class="loadingSpinner loadingOverlay-container"
      style="height:100px"
    >
      <div
        class="loadingOverlay optimizedCheckout-overlay"
      />
    </div>
    <div
      style="display:none"
    >
      <fieldset
        class="form-fieldset"
      >
        <legend
          class="form-legend is-srOnly"
        />
        <div
          class="form-body"
        >
          <div
            class="customerEmail-container"
          >
            <div
              class="customerEmail-body"
            >
              <div
                id="stripeupeLink"
              />
              <br />
              <div
                class="form-field"
              >
                <input
                  class="form-checkbox"
                  id="shouldSubscribe"
                  name="shouldSubscribe"
                  type="checkbox"
                  value="false"
                />
                <label
                  class="form-label optimizedCheckout-form-label"
                  for="shouldSubscribe"
                />
              </div>
            </div>
            <div
              class="form-actions customerEmail-action"
            >
              <button
                class="button stripeCustomerEmail-button button--primary optimizedCheckout-buttonPrimary"
                data-test="stripe-customer-continue-as-guest-button"
                disabled=""
                id="stripe-checkout-customer-continue"
                type="submit"
              />
            </div>
          </div>
          <p>
             
            <a
              data-test="customer-continue-button"
              id="checkout-customer-login"
            />
          </p>
        </div>
      </fieldset>
    </div>
  </div>,
  <div
    class="optimizedCheckout-form-input"
    id="stripe-card-component-field--input"
    placeholder="1111"
  >
    <div
      class="form-field--error"
    >
      <div
        class="optimizedCheckout-form-label"
        id="stripe-card-component-field--error"
      />
    </div>
    <div
      class="optimizedCheckout-form-label"
      id="stripe-card-component-field--label"
    />
  </div>,
]
`;
